### ISSUE TITLE
Add support for HDR gain maps

### ISSUE THREAD

**Issue by gregbenz** [ImageMagick/ImageMagick#6377]
### Is your feature request related to a problem? Please describe.

**HDR (high dynamic range)** is a significant leap forward in image quality on supporting monitors.

One of the challenges is to create an image which shows as an ideal HDR where possible, but safely falls back to showing SDR (standard dynamic range) where not supported. A new "**gain map**" standard (https://helpx.adobe.com/camera-raw/using/gain-map.html) offers this and is already supported by Chrome Canary v116 (including both MacOS / Windows desktop clients and the Android 14 beta). Enable the dev flag for gain maps via **chrome://flags/#gainmap-hdr-images**.

When images are reprocessed (such as by ImageMagick for posting on a Wordpress site via the media library), it is important that the gain map meta data and auxiliary image are retained. And if the image is resized/cropped, the secondary image should be altered in a consistent way.

When a JPG with a gain map is processed with ImageMagick now (ImageMagick 6.9.12-86 Q16 x86_64 17814), the primary image metadata (XMP-hdrgm) is retained, but the secondary image is discarded.

### Describe the solution you'd like

- Metadata and color profile of the base image are retained
- Auxiliary image is retained, including its metadata
- Auxiliary image is rescaled as needed to match any alteration of the primary image. This should be the same percent reduction (as the gain map itself may be at 1/2 or 1/4 resolution of the original to save space in the file).
- Solution should at a minimum support JPG gain maps encoded with an SDR base image (as this offers full backwards compatibility). It should ideally support AVIF gain maps (as this offers much smaller files with AVIF). Gain map support for JXL may be helpful, though the format lacks critical support to be important in the foreseeable future.

### Describe alternatives you've considered

Direct posting of the image without support for derivative assets like thumbnails or use of the media library. This could broadly affect a wide range of sites managing such images, so lack of support is very limiting and would require exporting multiple versions of images (if the platform even supports bypassing asset management which may reprocess the file).

### Additional context

See https://gregbenzphotography.com/hdr/ for examples, monitor tests, and more info on HDR images.

See https://helpx.adobe.com/camera-raw/using/gain-map.html for sample gain map images in JPG, AVIF, and JXL formats.

**Comment by snibgo** [comment]
I can't find any image files with gain maps on the Greg Benz site, or elsewhere. Can you point to some?

When an image has an associated gain map (GM), I can see that cropping and resizing the GM in parallel to the main image would be useful. For ImageMagick (IM), one difficulty is that the architecture is designed to process individual images, or lists of images. IM doesn't have a concept of "associated" images.

Another difficulty is: why limit parallel operations to cropping and resizing? How about general `-distort` operations? Sharpening? And many others?

Perhaps as a first step, IM could include an option to extract a gain map from a file, adding the gain map to its usual list of images. It could have another option to insert a gain map into a file. It would also read/write the GM metadata. The next step is the fairly simple image processing: given 2 out of 3 images (SDR, HDR, GM), make the third.

**Comment by gregbenz** [comment]
@snibgo There are some gain map examples under the Adobe link (direct download is https://www.adobe.com/go/gainmap_sample_photos).

Your question on other operations is a good one. I would say the support I outlined above would probably cover most needs (certainly the most critical ones for sharing images on the web). However, those additional operations would likely be useful to others (especially sharpening if some service wants to use that to offset softness caused by resizing). Personally, I would probably not invoke such options. My emphasis is simply on derivative versions of a finished image, especially for use with the WordPress media library.

I would imagine that internally extracting the gain map, scaling/cropping, retaining HDR metadata, and inserting back into the main image would work well. The scaling should be proportional rather than absolute pixels, as the specification allows for gain maps to be lower resolution than the primary image (they get scaled by the viewer to match the primary image, which ought to be a very good compromise in the interest of size reduction outside of some edge cases with high frequency detail in the HDR range). If the tool works that way internally, I would say it would be important that the tool automates that step rather than require the external management of that process to (1) simplify implementation, (2) ensure widespread adoption, and (3) avoid mistakes during 3rd-party implementation.

Auxiliary images seem to be showing up in other places, the iPhone appears to use it for at least some HEIC images: https://gist.github.com/kiding/fa4876ab4ddc797e3f18c71b3c2eeb3a. There are also examples of alpha channels like depth masks. To be clear, I don't personally see an obvious use case for HEIC / depth masks those with ImageMagick (as those are probably going to be used for inputs to editing tools like Photoshop), but just want to note there seems to be an increasing list of potential uses for associated images.

**Comment by snibgo** [comment]
Thanks for the samples.

IM has an option for extracting depth maps, "-define heic:depth-image=true", but it can't write depth maps into files, and doesn't treat then as "associated" images, to process in parallel with main images.

I suspect HDR will increase in popularity. The "gain map" system seems to be a useful bridge between SDR and HDR.

Parallel processing of associated images could be very messy. However, something like this might be feasible (bash syntax):
```
magick \
  -define hdr:read-gain-map=true \
  in.jpx
  -crop 2000x1000+100+200 +repage \
  -resize 600x600 \
  -define hdr:write-gain-map=true \
  out.jpx  
```
This would read images from `in.jpx`, including GMs. Each GM would be automatically resized to the same size as its parent image. `-crop` will crop all images (including GMs) to the same dimensions, then `-resize` will resize all images to fit in a 600x600 box. Finally, each image with its GM is written to `out.jpx`.

When IM reads the GM, it records any GM metadata as image properties. IM would also record the scale of the GM (eg 25%). When writing the GM, IM resizes the DM by the same scale, and writes the associated metadata.

Under this scheme, `-crop` and `-resize` etc would not need any modification. We would need changes to the coder for `jpx`, and coders for any filetype that might contain GMs.

How does that sound?

**Comment by fmw42** [comment]
If the gain map is pixel data (same size as input), then could it not be put into a meta channel?

**Comment by gregbenz** [comment]
> If the gain map is pixel data (same size as input), then could it not be put into a meta channel?

What's a beta channel?

One thing to note with the spec is that the child metadata varies by format, at least in the Adobe samples. In some cases (JPG if I recall correctly), critical gain map metadata is in the auxiliary image, so that probably needs to remain (not sure if the spec allows flexibility between that and the parent image metadata). There is always at least the gain map version in the parent (for quick parsing), but that may be it for some images.

**Comment by fmw42** [comment]
meta channel (not beta channel) is an extra channel apart from RGBA for example.  TIFF supports extra channels that it uses for example as masks. But they can be used for anything.  A depth channel could be stored in Imagemagick as a meta channel, for example.  I am not sure how it is stored now in HEIC images.

**Comment by snibgo** [comment]
fmw42 wrote:
> If the gain map is pixel data (same size as input), then could it not be put into a meta channel?

Ah, yes, thanks, I think that's a better idea. Using meta channels would mean the GM would have to be stored in memory at full scale, ie the same size as the main image.

The "Gain Map Specification (1.0 draft 12, May 2023)" https://helpx.adobe.com/content/dam/help/en/camera-raw/using/gain-map/jcr_content/root/content/flex/items/position/position-par/table/row-3u03dx0-column-4a63daf/download_section/download-1/Gain_Map_1_0d12.pdf says the GM will have either 1 or 3 channels. Hence there would be 1 or 3 meta channels.

IMagick is an interface between PHP and ImageMagick. (It isn't developed or supported by the same people as ImageMagick.) When ImageMagick has the capability to read and write GMs, IMagick could inherit that capability.

I don't have access to an HDR display. If I do the development work, I would rely on other people for testing.

**Comment by gregbenz** [comment]
@snibgo I would be happy to help test any gain maps you convert. You can reach me directly via gregbenzphotography.com/contact/

The Adobe AVIF images would be a also good option to test on an SDR system. They are built with an HDR base image, so they will show as tone mapped if gain maps are not supported and show the proper SDR image if the gain map is used. So if the converted AVIF matches the source AVIF on your screen, the gain map should be ok (the JPGs are built with an SDR base image and therefore would show the same result on your SDR display whether the gain map was valid or not). 

The Adobe materials include an image browser to view gain maps, which is a helpful reference.

**Comment by gregbenz** [comment]
> meta channel (not beta channel) is an extra channel apart from RGBA for example. TIFF supports extra channels that it uses for example as masks. But they can be used for anything. A depth channel could be stored in Imagemagick as a meta channel, for example. I am not sure how it is stored now in HEIC images.

Ah, makes sense. I'm used to referring to it as "alpha" channel (or "layer mask"), per Photoshop lingo.

**Comment by madmanchan** [comment]
Relevant CIPA document on Multi-Picture Format:  https://www.cipa.jp/std/documents/e/DC-X007_E.pdf
Note that while gain maps are an exciting direction for HDR imaging and formats, this is a swiftly moving area with many changes happening. There is an effort to standardize the approach, but that will take time. MPF is likely to be the mechanism used for JPEG (mostly for lack of a better option), but details like metadata and the math may change.

**Comment by gregbenz** [comment]
@snibgo Does jpeglib or the info from madmanchan cover what would be needed for IM to add support for JPG gain maps? (I can't offer much on your question on jpeglib, outside my expertise).

Anything I can do to help test?

**Comment by gregbenz** [comment]
JPG gain maps (aka "Ultra HDR JPG") are now supported in 4 browsers:
- mainline Chrome v116 (enable via chrome://flags/#gainmap-hdr-images).
- Brave v1.57.47 (via brave://flags/#gainmap-hdr-images)
- Opera v101 (via opera://flags/#gainmap-hdr-images)
- MS Edge Canary v117 (no dev flag needed)

They look amazing on all these browsers under a variety of SDR and HDR viewing conditions (and still look great as SDR on other older browsers that lack support since this format is automatically backwards compatible, the gain map data for HDR is just ignored when not understood or not applicable).

Sample gain maps: https://www.adobe.com/go/gainmap_sample_photos

**Comment by fmw42** [comment]
Gain maps are being researched but have not been implemented in IM.  The first step of supporting MPO JPGs has been implemented.  That is as far as I know has been done.  It is not an easy implementation for Gain Maps.

**Comment by gregbenz** [comment]
@fmw42 Thank you for the update! I'm sure there's a lot of detail to process two images in the same file.

**Comment by fmw42** [comment]
That part is implemented in IM to read the MPO JPG structure

**Comment by gregbenz** [comment]
@fmw42 Please forgive my ignorance here if this question sounds silly... If IM can now read and write MPO/MPF/auxiliary images, wouldn't the rest of the support for gain maps be fairly straight forward? It should be a matter of retaining the existing metadata and resizing/cropping the images.

Does the resizing cause edge artifacts with the interaction of the two images? I can imagine the encoding of the map itself might risk high frequency details, haven't dug into it. 

Or is there some other transformation being done? Conversions of color profile aren't needed (and should probably be avoided and just managed by the content creator).

**Comment by fmw42** [comment]
Sorry, I am not a developer, so do not know the code, only what I have heard from the developers discussions. So I could be wrong.  They would have to comment.

**Comment by gregbenz** [comment]
FYI that JPG gain maps are now supported in 4 browsers (representing ~75% of internet usage if all were at latest version - Safari support would bring this to 95%):

-     mainline Chrome v116 (via chrome://flags/#gainmap-hdr-images) - on desktop
-     Chrome Canary v118 (via chrome://flags/#gainmap-hdr-images) on Android 14 beta
-     Brave v1.57.47 (via brave://flags/#gainmap-hdr-images)
-     Opera v101 (via opera://flags/#gainmap-hdr-images)
-     MS Edge Canary v117 on Windows (no dev flag needed)

See https://www.adobe.com/go/gainmap_sample_photos for sample gain map images in Ultra HDR JPG format.

**Comment by gregbenz** [comment]
Chrome Canary v119 now includes support for AVIF gain maps as well.
Enable via chrome://flags/#avif-gainmap-hdr-images

**Comment by gregbenz** [comment]
Browser support for gain maps is now mainstream. Chrome v116+, MS Edge v116+, Brave v1.58+, and Opera v102+ all support JPG gain maps by default.

**Comment by fmw42** [comment]
What about Firefox and Safari?

**Comment by gregbenz** [comment]
I don't know if Safari 17 does, will test when MacOS releases next week. That would be the biggest missing piece (the others are 75% of browsing and Safari would get to 95% coverage). They added a lot of core gain map support which was highlighted in public sessions at WWDC.

FireFox share is very low and I'm not holding my breath for support soon, as I've seen no indications of progress.

The nice thing is that JPG gain maps fallback gracefully. So there is no penalty for sharing one of these images and you get the benefit if supported by both browser and the monitor.

Android 14 has support and looks awesome on a Pixel 7 Pro.

**Comment by martinschmidt-leica** [comment]
Hi @gregbenz , can you share a picture of Pixel 7 Pro that contains a gain map?

**Comment by gregbenz** [comment]
@martinschmidt-leica I'm confused, what's the purpose? Any JPG gain map will work on the Android 14 and the rendering is similar to what would show on a computer.

You can get a sample JPG gain map from the Adobe link above.

**Comment by martinschmidt-leica** [comment]
@gregbenz It's not about the map, but about the metadata. I would expect Google to implement Ultra HDR format according to the spec https://developer.android.com/guide/topics/media/platform/hdr-image-format. While the map itself is the same format, Android uses a slightly different container

**Comment by gregbenz** [comment]
@martinschmidt-leica thought you were asking for a picture "of" not created on the phone. 

I haven't explored enough, but haven't found a way to create one with the Android camera yet. I'm creating my images on a computer.

The Android and Adobe specs are above and should be identical.

Specific implementations have some degrees of freedom, such as the offset constants to avoid collisions with zero (choice of HDR base would be an option for formats other than JPG, but doesn't apply here as Android is addressing JPG). But that's all within the spec.

**Comment by gregbenz** [comment]
I've been told that JPG can be written with headers for both MPF and GContainer such that regardless of which is in the file, a reader could handle it if it only knows one or the other. That said, it isn't required or recommended in the spec as far as I know, so I don't believe you can assume you'll have that option if the goal is 100% compatibility.

**Comment by martinschmidt-leica** [comment]
FYI https://github.com/google/libultrahdr

**Comment by gregbenz** [comment]
@martinschmidt-leica After futher digging, my understanding is that an image can write both types of headers, that MPF is more common, and that when GContainer is used both headers are written. 

So as far as ImageMagick goes, it only needs to support MPF and will automatically support images which may include GContainer.

**Comment by gregbenz** [comment]
I've just posted a tutorial video and more details on how to create JPG gain maps using Adobe Camera RAW v15.5:
https://gregbenzphotography.com/hdr-images/jpg-hdr-gain-maps-in-adobe-camera-raw/

**Comment by gregbenz** [comment]
Android 14 was just released today and now officially supports HDR gain maps (Ultra HDR JPG) on Chrome, Edge, and Brave (Opera is not supported but works on desktop and uses same code base, so likely to follow soon). https://blog.google/products/android/android-14/

I've posted an image gallery with HDR gain maps at https://gregbenzphotography.com/hdr-gain-map-gallery/.

I also added a test (#7) to determine if an HDR-capable browser supports gain maps: https://gregbenzphotography.com/hdr/#tests

**Comment by gregbenz** [comment]
Adobe just released extensive HDR support to every version of **Lightroom**, including the ability to export HDR JPG gain maps from any device (Mac, PC, Android, iPad, and iPhone). https://gregbenzphotography.com/hdr-images/lightroom-adds-hdr-edit-and-export/

**Comment by maryla-uc** [comment]
> Chrome Canary v119 now includes support for AVIF gain maps as well. Enable via chrome://flags/#avif-gainmap-hdr-images

Hi, I would just like to point out that the gain map format for AVIF is still in flux, so current AVIF+gainmap images might not work in the future (actually, they most definitely won't). I'm glad to see there is interest for this, but make sure to treat it as a demo for now.

**Comment by gregbenz** [comment]
@maryla-uc Thanks for that info. To confirm, you mean gain maps for AVIF are in flux, you aren't saying JPG gain maps are, correct? As far as I know, JPG gain maps are stable and the Adobe-published proposal is what's in the ISO draft.

**Comment by gregbenz** [comment]
@maryla-uc Thanks. That has been my understanding, that JPG gain maps are set. They are supported in the Adobe standard by ACR, LR, Chrome, Brave, Edge, and Opera now in production release by default.

I have been under the impression that AVIF _might_ change and hence no encoder support and decoding is only under a dev flag.

**Comment by urban-warrior** [comment]
We are currently in the process of enhancing ImageMagick's support for HDR, but there's still a substantial amount of work ahead of us. Let's begin by addressing a fundamental feature: the ability to read JPEG images that contain both SDR (Standard Dynamic Range) and gain map images.

A straightforward approach to create an HDR image from the SDR and Gain map could involve the following command:

```
magick mpo:01.jpg -colorspace RGB -compose multiply -composite -define quantum:format=floating-point -colorspace sRGB 01.tiff
```

However, we are exploring the possibility of introducing a new composite operation, perhaps named `-compose gain-map`, to accurately convert SDR to HDR while considering the gain map metadata specified in the XMP profile.

We're seeking guidance from experts in Gain map handling to help define this new composite operation. What should the formulation look like, taking into account metadata parameters such as `hdrgm:GainMapMax` and `hdrgm:Gamma`?

Alternatively, if you have access to code-sets, preferably in C or C++, that demonstrate how to perform SDR to HDR conversion with gain maps, we would greatly appreciate it if you could point us in that direction. Once we've completed this initial step, we will also explore the possibility of converting HDR back to SDR while accounting for the gain map. Your expertise and guidance in this endeavor would be invaluable.

You can find a gain map specification plus sample images (including `01.jpg`) [here](https://helpx.adobe.com/camera-raw/using/gain-map.html).

**Comment by gregbenz** [comment]
@urban-warrior That's great to hear.

The issues you've just raised, if I understand correctly, would be expanded capability to generate an HDR with gain map from just an SDR input - correct? If so, perhaps a new issue should be raised for that (inverse tone mapping with gain map output), and then link back to this issue as a dependency. Or are you saying that you don't believe direct manipulation of the gain map would produce a quality result and that a temporary HDR result should be created in order to generate a new gain map?

--------
The intention of my original request here is simply to support keeping the image qualitatively the same when it is resized, compressed, or cropped. The goal being to support derivative files, such as thumbnails and small sizes in the WordPress media library. I don't believe that functionality would require any real understanding of the metadata or image content, but that it could serve as a foundation on which to build additional capability which does involve qualitative alteration of the content.

I believe the core functionality required to support the original request above includes:
- **Read and write metadata**. The tags may be new, but this should be straight forward. I see no need for the HDR-specific data to be altered here, simply passed through without change.
- **Read the auxiliary gain map** image in the MPF format (GContainer is written with MPF headers, so only MPF is needed). I don't know if this is supported or not. I know EXIFTOOL supports it and may be used to extract the gain map.
- **Write the auxiliary gain map** image in the MPF format (any tool which reads GContainer should support reading MPF). I don't know if this capability has been created to write depth maps or anything like that previously. I expect this is new and probably the core of effort required. The [libultrahdr](https://github.com/google/libultrahdr) library from Google may be a helpful resource here (see https://github.com/google/libultrahdr/blob/d1ad9a00076ec1ae593d0957f87a2d5921fc852d/multipictureformat.cpp#L30)
- If cropping / scaling the image, **the gain map should be scaled / cropped proportionally**.
- If cropping / scaling and the gain map is scaled using 1:4 or 1:2 (instead of a full 1:1), it must be scaled to 1:1, cropped/scale consistent with the base image, and then scaled down to the original 1:4 or 1:2 relationship originally used.

I don't believe this request requires ImageMagick to interpret HDR metadata, understand HDR, or construct an HDR image from the gain map. But perhaps I'm over-simplifying. I am unclear if the gamma encoding of the gain map falls outside normal bounds of processing SDR JPGs (ie, is there a quality benefit to converting any of the data to linear space for cropping/scaling, if that isn't already done when handling standard JPG now).

Please let me know if you need some image samples. I could create some gain maps with the same image output a few different ways to use as a reference to evaluate the impacts to the gain map when the same image is scaled or cropped.

Note that the Adobe Gain Map Demo App is very helpful for understanding gain maps conceptually. It represents the converted gain map (the RGB sampler does not list the encoded values, but rather the scalar encoded in linear space to get from SDR to HDR with the interpreted map). See the 12:10+ mark of this tutorial video: https://gregbenzphotography.com/hdr-images/jpg-hdr-gain-maps-in-adobe-camera-raw/

**Comment by arifdikic** [comment]
Hi All, we are happy to help adding ultra hdr support to imagemagick via libultrahdr. urban-warrior please let me know how we can sync up on the effort (libultrahdr originates from my team)).

**Comment by urban-warrior** [comment]
@arifdikic, At a minimum, if libultrahdr offers applications for encoding and decoding ultra HDR images, we could explore calling these apps from ImageMagick. Alternatively, the ideal scenario would be to compile libultrahdr on both Linux and Windows, enabling us to invoke its API from C. With thorough documentation, this integration would seamlessly incorporate libultrahdr into ImageMagick.

In the best-case scenario, your team could develop an ultrahdr coder specifically for ImageMagick, releasing it under the ImageMagick license. This would allow us to include it in the official ImageMagick distribution. What is the preference from your team?

To kickstart the process, could you share links to the libultra source, build instructions, API documentation, and an example codeset demonstrating the reading and writing of ultrahdr image files to/from disk?

**Comment by grapeot** [comment]
I happen to have a reference implementation based on Swift here: https://github.com/grapeot/AppleJPEGGainMap. It's still based on Apple's API, which doesn't have the greatest transparency, but hope it's somehow useful.

**Comment by gregbenz** [comment]
@grapeot I believe the Apple implementation is not compliant with the ISO spec (ISO 21496-1). Last I checked, it was using slightly different metadata and a 1D map (the ISO spec uses a 3D spec which should enable better color adaptation from HDR to SDR).
https://helpx.adobe.com/camera-raw/using/gain-map.html#resources

**Comment by arifdikic** [comment]
Hi @urban-warrior , 

Ultra HDR format is defined here: https://developer.android.com/guide/topics/media/platform/hdr-image-format 
(it's compatible with Adobe's format provided above, and we hope to standardize it soon as part of ISO and AOM efforts.) 

libultrahdr is located here: https://github.com/google/libultrahdr , it only depends on libjpegturbo. 

libultrahdr provides multiple encoding and decoding APIs as documented here: 

https://github.com/google/libultrahdr/blob/main/lib/jpegr.h

Now it's important to emphasize that, Ultra HDR is a dual intent format meaning that the image content to be encoded comes with an SDR-intent (read look) and an HDR-intent. The gainmap is normally generated from these two intents. 

Please take a look at the API surface for decoding and encoding to kick this off, and we can discuss the command line changes required and the integration work after clarifying any questions you might have. 
Again, my team is happy to give a hand for the actual integration work as well.

**Comment by arifdikic** [comment]
Quick update , we are looking into the code and working on a change proposal.

**Comment by gregbenz** [comment]
Sounds like JPG gain map support now is making its way to Instagram: https://www.theverge.com/2024/1/18/24043138/instagram-is-working-to-expand-hdr-photo-support-to-more-devices

**Comment by neckaros** [comment]
Does this request cover gain map support from HEIF/HEIC photos (mainly coming from iPhone)?Especially conversion from HEIF to Ultra JPG or Avif with gain map.

**Comment by gregbenz** [comment]
I recommend the emphasis be on formats which are on an ISO-standards track. Right now that is JPG gain maps. 

AVIF should follow the same path, but appears potentially subject to minor revisions before it may be more final / official. It's already supported in Chrome-based browsers (under a dev flag), so I would expect the risk for at least preliminary support (such as for encoding secondary images) won't be affected - I would expect it's more of a risk to things like metadata. When ready, this will be a very important format as it offers both smaller files and higher quality than JPG.

iOS HEIF have gain maps. I am unaware of a public spec, but the details seem straight-forward and your point on conversion accepting a HEIF source for conversion to another format like JPG is well taken. That would seem to be valuable to enable direct uploads of iPhone images to be shared in a format more widely supported by browsers (ie JPG gain map now - which would add HDR support for Chrome/Edge/Brave/Opera as well as SDR JPG fallback on FireFox vs failing to render anything at all).

**Comment by neckaros** [comment]
Agreed with you.

Most information I did found is under this gist which you already now about 😁
https://gist.github.com/kiding/fa4876ab4ddc797e3f18c71b3c2eeb3a

and the repo mentioned in the last comment
https://github.com/grapeot/AppleJPEGGainMap

**Comment by gregbenz** [comment]
@pmonck The Android GContainer approach adds an extra header but always includes MPF. Ideally, any reader would look for MPF in order to be universally compatible with any ISO-standards JPG gain map. 

If your test file is not supported, I would expect it is either because it looks for the GContainer header, or verifies the device hardware and rejects a source other than S24 (this might be based on metadata in the image or AndroidOS data about the host device which is being used to upload). Given promotion of S24 (no mention of S23 even though same OS), I wouldn't be suprised if we are in some exclusive marketing promotion window before IG opens support to upload from any device what can provide a compliant JPG gain map. This is just speculation on my part, but would make sense to me.

**Comment by arifdikic** [comment]
We are adding experimental support for HEIF/AVIF into libultrahdr based on the latest draft standard--> ImageMagick should get these out of box as well. It will also going to be transcode in between these e.g. HEIF w gainmap --> AVIF 10-bit HDR etc.

**Comment by arifdikic** [comment]
ImageMagick changes are to be shared soon for review. Who'd review the API changes required?

**Comment by martinschmidt-leica** [comment]
> We are adding experimental support for HEIF/AVIF into libultrahdr based on the latest draft standard--> ImageMagick should get these out of box as well. It will also going to be transcode in between these e.g. HEIF w gainmap --> AVIF 10-bit HDR etc.

That's great news! Are there also plans for JPEGXL support?

**Comment by gregbenz** [comment]
@arifdikic If that's gain map support for AVIF (and HEIF), that's incredible news. The sample AVIF images posted by Adobe look incredible in Chome (with dev flag enabled). It will be incredibly valuable to enable use of AVIF as a gain map format given the substantial file size reduction and quality gains it offers.

**Comment by DichenZhang1** [comment]
Hi all,

We have sent out the [pull request](https://github.com/ImageMagick/ImageMagick/pull/7198) for ImageMagick and libultrahdr integration. There are still some open discussions, but the basic encoding, transcoding, editing functions were tested working well.

Thank you!

**Comment by arifdikic** [comment]
Dichen, you can also share the libultrahdr2.0 roadmap here , which should come to ImageMagick once it's out.

**Comment by gregbenz** [comment]
@AngelaDMerkel Gain maps and HDR image formats are actually two different mechanisms to deliver HDR, and they can be used together or independently of each other.

There is a draft **ISO HDR spec**, and this serves to provide an HDR and basic metadata which could be used to tone map the image to less capable displays. Generally, you'd want 10-12 bit support to use such a format (to avoid banding). JXL supports this.

Then there is the draft **ISO gain map spec**, and provides two images in the same file (a base image and the actual gain map) and some metadata which are used to construct an SDR, a full HDR, or something in between. The gain map can convert an SDR base to HDR or an HDR base to SDR, either is possible. If the base image is encoded as HDR, then it should ideally use the ISO standard for that. If the base image is encoded as SDR, then that HDR spec is not relevant (because tone mapping is not used with gain maps, which is a huge benefit because it allows higher quality and avoids variability in rendering across different software). JXL supports gain maps, and Adobe has provided both sample JXL gain maps and a demo app which can view them (no browser supports this yet as far as I know). https://helpx.adobe.com/camera-raw/using/gain-map.html

==========

The most important gain map format right now is **JPG**. This is a de facto standard and widely supported by Adobe, Android, and most web browsers. It's the only reliable way to share high-quality HDR on the web right now in a way that still looks great on less capable displays. 

The next most important file type to support is **AVIF gain maps**. It is set to become the dominant way to share HDR on the web as it offers higher quality at much smaller file sizes. Adobe has provided samples of AVIF gain maps (same link above) and Chrome supports them if you enable the dev flag (chrome://flags/#avif-gainmap-hdr-images). Support for this format is extremely important, and comments in the libultrahdr repo suggest support for it is coming (though I'm not sure the implementation details - does it support both SDR and HDR base, RGB gain maps or luminosity only, etc).

**JXL** would be an excellent gain map candidate too, but there is minimal browser support for even JXL without a gain map - so that is a long ways off at best. As a JXL gain map standard is not finalized and there isn't a reference implementation for a browser, it seems a bit premature to me to add support. _Do you see some use case you see where a JXL gain map would important in the near term?_

JXL has some benefits over AVIF, but I expect the visual quality should be indistinguishable for most HDR web use. Progressive decoding is a potentially relevant benefit of JXL, but I would be surprised if it works in the context of a gain map. Speed might be a benefit in some edge cases, but I suspect ongoing efforts to add AV1 encoding support in hardware. Both support up to lossless encoding, so differences there come down to size vs quality for specific images. I have read some reports of AVIF size limits at 4k, which are not correct (I've tested AVIF and seen that Adobe software supports at least 20,000px square and Chrome supports at least 10,000 square - I didn't bother to find the ultimate limits). Overall, they are probably not that different to the average user and AVIF universal support with modern browsers that make it the clear choice for the foreseeable future of the web.

JXL has some nice benefits in other contexts (32-bit support for subsequent editing, direct conversion of existing JPGs without further loss of quality, etc), but I don't believe those really matter in scenarios where you would use a gain map.

==========

FYI that the Adobe spec above covers details for potential gain maps in many other formats (**HEIC, TIF, DNG, PNG**). The iPhone already captures HEIC gain maps, but there is no published spec and I believe is different from the Adobe proposal (though I haven't spent much time reviewing those details as browser support for HEIC is limited to Safari). libultrahdr comments also note intended support for HEIC, which would be very handy if that means support to decode the Apple format in order to transcode to more widely supported formats like the ISO JPG gain map. That would be very helpful for iPhone users to be able to upload images directly from their phone to various websites, for example.

**Comment by AngelaDMerkel** [comment]
> @AngelaDMerkel If your goal is simply to use a JXL HDR to produce a gain map in another format (JPG, AVIF, etc), then that should be possible when the libultrahdr implementation is ready 

This would be my use case more often than not. As you describe, I would like to retain JXL HDR images as the archival masters and convert to AVIF HDR for thumbnails and perhaps JPEG HDR for full resolution images or whatever the preference is. Ideally, I would prefer this feature to be bi-directional: that is, converting from JXL HDR to JPEG Ultra HDR, and back again. I would prefer to encode full fat HDR in what I assume is [this spec you've described](https://www.color.org/hdr/07-Nicolas_Bonnier.pdf)? 

> Do you see some use case you see where a JXL gain map would important in the near term?

Unless there is more web support for JXL, I don't see a very likely use case. I have moved over to JXL for all my internal work, but me and my peers are using colour managed software and wide gamut monitors, and not really targeting web use for what we are doing. If JXL were more widely supported I would use that because of progressive decode, but it's not in use (widely). 

Thank you for the clear explanation here. If libultrahdr does indeed introduce some support for JXL HDR base image reencoding, all the tools I will need will be there. If there is at least PPM support it'll be workable.

**Comment by gregbenz** [comment]
@AngelaDMerkel I would suspect (but have not tested to confirm) that your goals will supported with this integration. libultrahdr is offering a way to encode and decode images and ImageMagick supports JXL. That's not to say there might be implementation details that result in bugs that need to be addressed, so I would test when ready.

That document you reference is the reference for ISO HDR (gain maps are a different ISO draft spec). It's not something you should need to think about if you aren't writing software. For example, exports from Adobe software support it automatically. But in practice, you might want to be aware of it to make sure 3rd-party transcoding doesn't strip that metadata (ie, a web service might degrade the quality of your image if this is removed). A lot of sites still strip color profiles, so we might see some mistakes as HDR and gain map data roll out.

I would still probably avoid JXL in relation to gain maps at this point. Converting a simple HDR image to a gain map offers little benefit (might help in a few cases if that's all a given service will only accept a gain map image for HDR content, but offers no quality improvement if you aren't supplying the SDR). And if you're trying to use a gain map provided to you as an HDR source, you can simply open it directly in Adobe software. Of course, you may use other editors that require some other intermediate file, but I'd probably just convert to something like TIF because it's probably faster than JXL encoding for a temporary file.

**Comment by eddy-geek** [comment]
From the opening post of this long but productive thread:

> When a JPG with a gain map is processed with ImageMagick now (ImageMagick 6.9.12-86 Q16 x86_64 17814), the primary image metadata (XMP-hdrgm) is retained, but the secondary image is discarded.

What is the new state of this after #7198 and others ? My use-case is the simplest one, simply apply to the main image and keep the secondary gain map as-is, with e.g.:

```sh
magick mogrify -format jpg -quality 75 *.jpg
```

1. can a build of new ImageMagick *without* libultrahdr, e.g. latest (7.1.1-31) AppImage retain the secondary ?
2. can a build *with* libultrahdr do it ? does it require special command-line flags ?

For (1) my tests indicate it does not:

<details><summary>to reproduce on linux...</summary>
<p>
```sh
wget https://github.com/MishaalRahmanGH/Ultra_HDR_Samples/raw/main/Originals/Ultra_HDR_Samples_Originals_01.jpg
wget https://github.com/ImageMagick/ImageMagick/releases/download/7.1.1-31/ImageMagick-2f6d2de-clang-x86_64.AppImage
chmod +x ImageMagick-2f6d2de-clang-x86_64.appimage
exiftool Ultra_HDR_Samples_Originals_01.jpg | grep -w MP
./ImageMagick-2f6d2de-clang-x86_64.appimage mogrify -format jpg -quality 75 Ultra_HDR_Samples_Originals_01.jpg
exiftool Ultra_HDR_Samples_Originals_01.jpg | grep -w MP
```

the first `exiftool`  command display the  following, while the second displays nothing:

> MP Image Flags                  : (none)
> MP Image Format                 : JPEG
> MP Image Type                   : Undefined
> MP Image Length                 : 62570
> MP Image Start                  : 2684148
> MP Image 2                      : (Binary data 62570 bytes, use -b option to extract)
</p>
</details> 

could it be considered, as distro/appimage support for libultrahdr will probably take more time ?

*(thanks a lot for this great work!)*

**Comment by eddy-geek** [comment]
for point (2) above I think [this](https://github.com/ImageMagick/ImageMagick/pull/7198#issuecomment-2028071428) comment (related to this [commit](https://github.com/ImageMagick/ImageMagick/commit/2d47425e9002a370a5921b2e609d52e755162769)) answers me at least partially ;

> We've implemented several patches to enhance support for UHDR in ImageMagick. Currently, we've removed the UHDR calls from JPEG since they wouldn't function properly anyway, and we don't advocate direct calls from one coder to another. **Instead, for any UHDR disguised as JPEG, use udhr:myImage.jpeg.**

**Comment by isenberg** [comment]
It's available now at least for UltraHDR JPEGs with gain maps I saw and tried out this week: #7198

And thanks to @gregbenz, your page is a great start into the HDR display topic!

**Comment by radogado** [comment]
Just noticed Apple Photos for Mac exports JPEG gain map HDR, but Safari doesn't support it and displays SDR only 🙄

**Comment by urban-warrior** [comment]
Can you post a link to an example JPEG with gain map here so we can download and verify ImageMagick works well with the image and captures the correct metadata?

**Comment by gregbenz** [comment]
@urban-warrior I have several on my site:

https://gregbenzphotography.com/hdr/#tests

https://gregbenzphotography.com/hdr-gain-map-gallery/

The tests section of the first link includes multiple encodings (Adobe JPG, ISO JPG - and JXL, the AVIF gain map is actually non-standard and needs to be reposted). The gallery link currently uses the Adobe format. 

Can also send you something using the Google JPG encoding, which is probably best as that is the current capability of libultrahdr (ISO is the future and would be a natural update down the road, which should be transparent to IM other than updating the dependency when available). I would expect it should probably transcode a source using the Adobe format, but have not tried using it to decode it yet.

Please let me know if there is something specific you'd like to have.

**Comment by gregbenz** [comment]
@martinimal libultra has a constant flag which in theory would enable ISO encoding (same one you pointed out), however the output is not currently valid. It will show as SDR only in Adobe software and will actually fail to even render the base SDR image in Chrome if you turn off the backup XMP encoding. 

That ISO support flag was added a while ago, and it is likely that something simply changed in the draft spec after the libultra code was written. It's probably just some minor update to libultra to get it updated to the current ISO draft standard. I've reported the issue for review.

**Comment by martinimal** [comment]
> @martinimal libultra has a constant flag which in theory would enable ISO encoding (same one you pointed out), however the output is not currently valid. It will show as SDR only in Adobe software and will actually fail to even render the base SDR image in Chrome if you turn off the backup XMP encoding.
> 
> That ISO support flag was added a while ago, and it is likely that something simply changed in the draft spec after the libultra code was written. It's probably just some minor update to libultra to get it updated to the current ISO draft standard. I've reported the issue for review.

@gregbenz, understand. Did you check the updates from the last days? I enabled both flags in the latest version and was able to create images that are HDR in browser and in Adobe app. I use API4 to asssemble jpeg and gain map.
I did not try to disable XMP meta data.

**Comment by gregbenz** [comment]
@martinimal The updates allow proper ISO encoding, with one outstanding issue which will cause both XMP and ISO encoding to throw an error. A [fix](https://github.com/google/libultrahdr/pull/315) has been created and works great in my testing. Once that is merged to mainline, it appears to be in great shape. 

I've tested exports with ISO + XMP encoding as well as ISO only. They both generally work great. I have seen some errors trying to view the output, but these are problems with other software not libultra.

There are scenarios where disabling XMP will be an issue, so I recommend keeping it for now until ISO support has expanded a bit further. Skipping the XMP saves ~1K in file size and is probably ideal in the long term, but not worth it just yet.

**Comment by Devkiller41069** [comment]
Fu*k**g hackers

**Comment by Devkiller41069** [comment]
👀

**Comment by isenberg** [comment]
I'm already using IM 7 now with libuhdr 1.3.0 and it produces gainmap jpegs viewable in Opera stable and Chrome 130 stable and 132 beta. But I failed so far to generate gainmap jpegs showing up in macOS Preview.app with a different color space than the combination of Rec.2020 (bt2100) with the PQ transfer function. Maybe that's a bug in Preview.app as my images work fine in any browser just reading the SDR part so they must actuall contain data with the HLG transfer function.

Command used:
`magick -define uhdr:sdr-color-gamut=bt709 -define uhdr:hdr-color-gamut=bt709 -define uhdr:hdr-color-transfer=hlg -define uhdr:gainmap-quality=90 sdr.tif hdr.tif -quality 95 uhdr:output.jpg`

Just for a test to check if the parameters I read: When replaying the input parameters from bt709 to bt2100 I see changes towards slightly wrong colors as expected as my input data is sRGB. But independenlty from hdr-color-transfer=pq or hdr-color-transfer=hlt macOS Preview.app always shows PQ and Rec.2020.

Or is Rec.2020 absolutely required when using HLG without any chance to use displayP3?

**Comment by gregbenz** [comment]
@isenberg If the output shows as ISO HDR in the Adobe Gain Map Demo app (use ctrl/cmd-I to show the info overlay), then the output should be valid ISO encoding and should work in Apple software. That app is the most reliable way to confirm valid output.

That said, I have seen a handful of HDR bugs in the recent Apple updates (iMessage only shows HDR on the local end of the message, MacOS Photos shows SDR even though the Photos app is fine on iOS/iPadOS, etc). And Apple software prior to the latest updates had several color-space specific bugs (sRGB tended to be most prone to issues).

If you have a valid ISO gain map that works elsewhere but not in Apple software, I would log a bug with your sample image (per https://feedbackassistant.apple.com/).

Question on your commands, do you need to specify the color gamuts in the IM command? Ideally, the primaries could be assumed from common ICC profile names and passed to uhdr, and the output EOTF should just default to HLG if not specified (that's what the standalone ultrahdr_app does).

**Comment by isenberg** [comment]
I suspect some issue combined with the type of gainmap JPEG my scripts produce and what the iOS API expects I use in my own application to display in SwiftUI. The Adobe example images and app I'll try for comparison. Currently working on generating gainmp JPEGs from Rec.2020 input files instead to see if that helps as I anyway planned to replace sRGB as image source. I'm generating those images myself from raw camera data so can decide about the color space, just used sRGB in the beginning as it looked like less problematic.

**Comment by gregbenz** [comment]
@isenberg Feel free to post an example here with an image encoded via IM that you don't see working properly in Apple Preview (would help to note the specific IM build and encode parameters for the image).

And is latest IM pulling libultra 1.3 with ISO enabled by default now, or are you building from a modified version of either IM or libultra?

**Comment by isenberg** [comment]
Thanks Greg! The Gain Map Demo App you suggested, got it from https://helpx.adobe.com/camera-raw/using/gain-map.html , solved my question as there I could see in the info overlay now for my images:
Use Base Color Space: yes
Base Color Space: sRGB IEC61966-2.1
Alternate Color Space: Unspecified (same as base)
Gain Map Type: Adobe / Ultra HDR
Gain Map Weight: 1.0

While the same image in Preview.app shows BT.2020 Primaries; PQ (Gain Map Preview ...)

Looks like the Preview.app has a bug about that info text. Otherwise the image looks visually identical on the HDR screen.

With the gain map demo app's channel histogram graph I could also see why some of my images look more red in HDR: The integral below the red curve increases with higher HDR headroom and as the transfer curve is logarithmic in HDR that could explain why the visual ratio of red increases. It's about the images on my website areo.info/mars20 where you can see the effect of more red in some when switching between SDR and HDR display. 

Let's see what my experiments with Rec.2020 soon show...

**Comment by isenberg** [comment]
I used ImageMagick 7.1.1-41 and libultrahdr 1.3.0.
The first tests now with Rec.2020 looks usable, also work as SDR in other browsers. If the red effect disappears I don't know yet as I have to adjust my profile now for Rec.2020 as I use different tools for the image production with that colorspace, but as I anyway have to move to that, will stay on Rec.2020 now.

About the Preview.app I'll file a bug at Apple.

**Comment by isenberg** [comment]
And one note about the Gain Map Demo app: I now see the detail it writes "Use Base Color Space: Linear ProPhoto RGB" in the info layer, while correct "Rec2020 Gamut with sRGB Transfer + HDR + Gainmap" below the image.

I guess that should be corrected to "Used Base Color Space" as that's the space used in the tool before the image was written as Rec2020. Or is the HDR display really switching now to ProPhoto linear? I don't think so as the image looks visually identical in the other tool where I created it with and in the demo gain map app.

**Comment by gregbenz** [comment]
@isenberg With the ISO draft being unpublished, hard to say what the exact expectations are for the alternative HDR color space. But logically, you'd assume the base is encoded as stated, then the HDR numbers are derived from it and the numbers get interpreted using specified alt (HDR) color primaries, or under the base (SDR) primaries if not explicitly stated. 

You cannot assess a ProPhoto or Rec2020 image visually in a reliable way as both vastly exceed the limits of even wide gamut monitors (ProPhoto can even encode values outside the range of human vision). P3 is probably a better choice (especially for JPG gain maps), but if you use Rec2020, you'd want to validate the numbers rather than assess visually.

I'll contact you separately, as I believe the questions you raise are probably unrelated to IM itself. If there is any color error here, it is more likely in the source image, encoding parameters, the dependent library (libultra), decoding software for analysis, or the OS / environment (which would affect any visual assessment).

**Comment by isenberg** [comment]
I agree, P3 would make more sense, especially as that's also almost matching iOS devices capabilities. I guess I started experiments with Rec.2020 as I got the impression that is more common for HLG HDRs than P3. Then let's continue privately and we can later share results here if usable for the public.

**Comment by gregbenz** [comment]
Just to close out the discussion on the past few comments... I believe things are working as expected. After reviewing the image sample from @isenberg, I believe the issue noted is unrelated to IM. The current version of libultra does not explicitly support a mix of different color spaces between the base (SDR) and alternate (HDR) image. I suspect libultra is not designed to offer any color conversion and therefore this use of mixed primaries in the input is not supported. Conversion is not necessary as a better solution is pending.

There is an open pull request which would explicitly support this use case. So keeping base/alternate in the same space works now, and a mix of spaces should be supported with an update to libultra in the near future. Should not be an issue by the time libultra becomes included with ImageMagick by default.

**Comment by radogado** [comment]
I was unable to use JPEGs with HDR gain maps from Mac Photos to WordPress, no matter what, so I did the following workaround:

- Upload the exported HDR JPEGs to WP, letting it strip HDR from the downsized versions.
- Note the downsized WP images' resolutions
- Back in Mac Photos, export multiple times per each WP resolution
- Rename each group to the WP naming convention, e.g. IMG_4910-2048x1152.jpeg. Automated with a script from ChatGPT, because they are 500 photos.
- Upload to WP via FTP and overwrite the files

Finally this works (Chrome only) , but I'm looking forward to a better workflow.

**Comment by jishi92** [comment]
Hi,all.I currently need to support UHDR JPG encoding and decoding in ImageMagick, but I'm not sure how to handle geometric operations like scaling and cropping for the base image and gainmap. Should the gainmap be placed in the image structure itself? Or should it be treated as the next node of the base image? However, jpeg only returns the first frame. Do you have any good suggestions? Additionally, when will the open source community support UHDR images?

**Comment by gregbenz** [comment]
>  Additionally, when will the open source community support UHDR images?
It already does, including ImageMagick. More info and other projects (the key library is libultrahdr, which is supported by Google): https://gregbenzphotography.com/hdr/#developers

**Comment by jishi92** [comment]
> > Additionally, when will the open source community support UHDR images?
> > It already does, including ImageMagick. More info and other projects (the key library is libultrahdr, which is supported by Google): https://gregbenzphotography.com/hdr/#developers
<think>

I found the ReadUHDRImage method in the source code, indicating that it already supports decoding of uhdr images. However, I haven't discovered any scaling processing related to gainmap. For example, when converting a uhdr jpeg to a uhdr heif, I'm unaware of how to handle scaling and cropping operations for gainmap. Do you have any example code I can refer to?

**Comment by gregbenz** [comment]
@jishi92 Should be the standard commands for cropping / scaling: https://usage.imagemagick.org/crop/

@ram-mohan Anything that would be unique for using IM with libultra to crop or resize a gain map?

**Comment by jishi92** [comment]
> [@jishi92](https://github.com/jishi92) Should be the standard commands for cropping / scaling: https://usage.imagemagick.org/crop/
> 
> [@ram-mohan](https://github.com/ram-mohan) Anything that would be unique for using IM with libultra to crop or resize a gain map?

Generally speaking, the width and height of the gainmap and baseimage are not consistent, so scaling factors need to be stored for synchronized processing.

**Comment by gregbenz** [comment]
@jishi92 The dimensions are very often identical (for example this is what Adobe Lightroom exports, a 1:1 map). 

Smart phones may use 1:2 maps to save space and other reduced resolution maps are possible. It is always a fixed reduction ratio using an integer. The height and width cannot scale to arbitrary or independent sizes.

When transcoding, the map scaling factor should be kept the same, but down-sampling could be done to further shrink the size of the image. However, I would recommend keeping the same scaling factor unless you know what you are doing and are testing the results to ensure you won't create artifacts.

The scaling factor is managed in libultra, which is the dependency used by ImageMagick to transcode gain maps (It is kMapDimensionScaleFactor in the code base, and would be -s # in their command line demo app, where the # is the scale reduction in the map, so 1 would be the highest quality 1:1).

I recommend you read the ISO spec if you have questions on these factors. Links for that and other resources here: https://gregbenzphotography.com/hdr/#developers

**Comment by jishi92** [comment]
> [@jishi92](https://github.com/jishi92) The dimensions are very often identical (for example this is what Adobe Lightroom exports, a 1:1 map).
> 
> Smart phones may use 1:2 maps to save space and other reduced resolution maps are possible. It is always a fixed reduction ratio using an integer. The height and width cannot scale to arbitrary or independent sizes.
> 
> When transcoding, the map scaling factor should be kept the same, but down-sampling could be done to further shrink the size of the image. However, I would recommend keeping the same scaling factor unless you know what you are doing and are testing the results to ensure you won't create artifacts.
> 
> The scaling factor is managed in libultra, which is the dependency used by ImageMagick to transcode gain maps (It is kMapDimensionScaleFactor in the code base, and would be -s # in their command line demo app, where the # is the scale reduction in the map, so 1 would be the highest quality 1:1).
> 
> I recommend you read the ISO spec if you have questions on these factors. Links for that and other resources here: https://gregbenzphotography.com/hdr/#developers

thank you gregbenz,I find kMapDimensionScaleFactor

**Comment by ram-mohan** [comment]
> I found the ReadUHDRImage method in the source code, indicating that it already supports decoding of uhdr images. However, I haven't discovered any scaling processing related to gainmap. For example, when converting a uhdr jpeg to a uhdr heif, I'm unaware of how to handle scaling and cropping operations for gainmap. Do you have any example code I can refer to?

Currently the uhdr library supports uhdr jpeg format only. uhdr heif and uhdr avif are in experimental branch and not mainlined yet. So in IM, if you intend to transcode uhdr jpeg to uhdr heif, currently it is not possible. On the other uhdr jpeg to heif is possible where uhdr jpeg is first decoded and any further image processing operations are applied (like resize/crop/rotate/...) and then encoded to heif. Note that this heif is not uhdr heif but a regular heif file with out any gainmap.

**Comment by ram-mohan** [comment]
> > [@jishi92](https://github.com/jishi92) Should be the standard commands for cropping / scaling: https://usage.imagemagick.org/crop/
> > [@ram-mohan](https://github.com/ram-mohan) Anything that would be unique for using IM with libultra to crop or resize a gain map?
> 
> Generally speaking, the width and height of the gainmap and baseimage are not consistent, so scaling factors need to be stored for synchronized processing.

If the uhdr image needs to be resized to a new dimension, one can imagine that, basing on the primary image dimension and target image dimension, the resize factor needs to be computed along horizontal and vertical direction and then this factor needs to be applied to primary image and gainmap image (that are presented in uhdr image) separately. The resized primary image and gainmap image are combined to a final uhdr image. 

But in IM this is not what happens, if the uhdr image is intended to be resized,  it is first decoded. Decoding a uhdr image would result in only one image (primary image with gainmap applied) and not two images (primary and gainmap images separately). This decoded image gets resized. This resized image again when encoded to uhdr format, as only one image is presented as input, uhdr library performs tonemapping internally, computes gainmap, encodes tonemapped image and gainmap image to a uhdr image. The resulting primary image (tonemapped) and gainmap image may not be the same as the ones we would get had we resized the original primary and gainmap image

**Comment by gregbenz** [comment]
Now that JPEG gain-map support has landed, I’m closing this issue as completed.

I’ve opened #8916  to track AVIF (and possible HEIF) gain-map support, which are the next logical steps and would be more easily discussed in a new thread.

**Comment by arpu** [comment]
@gregbenz can you ppoint me to the doc to generate jpg with gainmap from hdr file?

**Comment by gregbenz** [comment]
@arpu You need to provide both the SDR and HDR input to create a useful gain map. Direct conversion of an HDR image to a gain map is nearly always going to produce a low-quality result (though the failure may not be obvious until you view on an SDR display). You're better off just sharing the HDR as is if you don't have a map or SDR version to create it.
